README.md
# Ohjelmistotekniikka, harjoitustyö

[työaikakirjanpito](https://github.com/roni-b/Ohjelmistotekniikka-2022/blob/master/dokumentaatio/tyoaikakirjanpito.md)

[vaatimusmäärittely](https://github.com/roni-b/Ohjelmistotekniikka-2022/blob/master/dokumentaatio/vaatimusmaarittely.md)

[changelog](https://github.com/roni-b/Ohjelmistotekniikka-2022/blob/master/dokumentaatio/changelog.md)

[arkkitehtuuri](https://github.com/roni-b/Ohjelmistotekniikka-2022/blob/master/dokumentaatio/arkkitehtuuri.md)


dokumentaatio/changelog.md
## Viikko 3

- Tehty pelille ikkuna ja ruudukko sekä pääsilmukka
- Lisätty luokka joka vastaa uuden palikan putoamisesta ja sen liikuttamisesta nuolinäppäimillä

## Viikko 4

- Lisätty luokka joka vastaa jo pudotettujen palikoiden säilyttämisestä
- Lisätty myös taustalle ja ruudukolle oma luokka
- Lisätty funktio törmäysten tunnistamista varten
- Lisätty metodi palikan kääntämistä varten sekä metodi joka antaa uuden palikan


src/tetris.py
import pygame, random


BLOCK_SIZE = 50
height = 20
width = 10

#grid
grid = [pygame.Rect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE) for x in range(width) for y in range(height)]

from pygame.locals import (
        K_DOWN,
        K_LEFT,
        K_RIGHT,
        K_UP
    )

class NewShape(pygame.sprite.Sprite):
    def __init__(self, picture_path):
        super(NewShape, self).__init__()
        #make surface
        self.surf = pygame.image.load(picture_path)
        #draw rect around
        self.rect = self.surf.get_rect()

    def update(self, pressed_keys):
        if pressed_keys[K_DOWN]:
            self.rect.move_ip(0, 1)
        if pressed_keys[K_LEFT]:
            self.rect.move_ip(-1, 0)
        if pressed_keys[K_RIGHT]:
            self.rect.move_ip(1, 0)
        if pressed_keys[K_UP]:
            self.rotate()

        if self.rect.left < 0:
            self.rect.left = 0
        if self.rect.right > display_width:
            self.rect.right = display_width
        if self.rect.top <= 0:
            self.rect.top = 0
        if self.rect.bottom >= display_height:
            self.rect.bottom = display_height
   
    def rotate(self):
        self.surf = pygame.transform.rotate(self.surf, 90)

display_height = height * BLOCK_SIZE
display_width = width * BLOCK_SIZE

display = pygame.display.set_mode((display_width, display_height))
pygame.mouse.set_visible(False)







#specify object
next_shape = random.randint(1, 7)
print(next_shape)
shape = NewShape("shapes/"+str(next_shape)+".png")
#group containing sprites and then drawing them
shape_group = pygame.sprite.Group()
shape_group.add(shape)
pygame.init()




running = True

while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    pressed_keys = pygame.key.get_pressed()
    shape.update(pressed_keys)
    display.fill((0, 0, 0))
    [pygame.draw.rect(display, (40, 40, 40), i_rect, 1) for i_rect in grid]

    for entity in shape_group:
        display.blit(entity.surf, entity.rect)
    

    pygame.display.update()
pygame.quit()



src/tetris_main.py
import os
import random
import pygame

wd = os.getcwd()
clock = pygame.time.Clock()

BLOCK_SIZE = 50
HEIGHT = 20
WIDTH = 10
DISPLAY_HEIGHT = BLOCK_SIZE * HEIGHT
DISPLAY_WIDTH = BLOCK_SIZE * WIDTH

display = pygame.display.set_mode((DISPLAY_WIDTH, DISPLAY_HEIGHT))
pygame.mouse.set_visible(False)


grid = [pygame.Rect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE) for x in range(WIDTH) for y in range(HEIGHT)]

from pygame.locals import K_DOWN, K_LEFT, K_RIGHT, K_UP


class NewShape(pygame.sprite.Sprite):
    def __init__(self):
        super().__init__()
        next_shape = random.randint(1, 7)
        self.image = pygame.image.load(os.path.join(os.path.dirname(__file__), 'shapes', str(next_shape)+'.png')).convert_alpha()
        self.image = pygame.transform.scale(self.image, (150, 105))
        self.rect = self.image.get_rect()

    def input(self, pressed_keys):
        if pressed_keys[K_DOWN] and self.rect.bottom < DISPLAY_HEIGHT and collision() is False:
            self.rect.move_ip(0, 5)
        if pressed_keys[K_LEFT] and self.rect.bottom < DISPLAY_HEIGHT and collision() is False:
            self.rect.move_ip(-5, 0)
        if pressed_keys[K_RIGHT] and self.rect.bottom < DISPLAY_HEIGHT and collision() is False:
            self.rect.move_ip(5, 0)
        if pressed_keys[K_UP] and self.rect.bottom <= DISPLAY_HEIGHT:
            self.rotate(self.image.get_width(), self.image.get_height())

        if self.rect.left < 0:
            self.rect.left = 0
        if self.rect.right > DISPLAY_WIDTH:
            self.rect.right = DISPLAY_WIDTH
        if self.rect.top <= 0:
            self.rect.top = 0
        if self.rect.bottom >= DISPLAY_HEIGHT:
            self.rect.bottom = DISPLAY_HEIGHT
            shapes.add(DroppedShapes(self.image, self.rect))
            self.give_new_shape()
    def rotate(self, rect_x, rect_y):
        self.image = pygame.transform.rotate(self.image, 90)
        self.rect = self.image.get_rect(center = self.image.get_rect(center = (rect_x, rect_y)).center)

    def drop(self):
        if not collision():
            self.rect.move_ip(0, 1)
        else:
            self.rect.move_ip(0, -3)
            shapes.add(DroppedShapes(self.image, self.rect))
            self.give_new_shape()


    def update(self, pressed_keys):
        self.input(pressed_keys)
        self.drop()

    def give_new_shape(self):
        next_shape = random.randint(1, 7)
        self.image = pygame.image.load(os.path.join(os.path.dirname(__file__), 'shapes', str(next_shape)+'.png')).convert_alpha()
        self.image = pygame.transform.scale(self.image, (150, 105))
        self.rect = self.image.get_rect()

class DroppedShapes(pygame.sprite.Sprite):
    def __init__(self, image, rect):
        super().__init__()
        self.image = image
        self.rect = rect

class Background(pygame.sprite.Sprite):
    def __init__(self):
        super().__init__()
        self.image = pygame.Surface([600,1200])
        self.image.fill((255, 255, 255))
        self.rect = self.image.get_rect()

class Grid(pygame.sprite.Sprite):
    def __init__(self):
        super().__init__()
        

shape = pygame.sprite.GroupSingle()
shape.add(NewShape())

shapes = pygame.sprite.Group()

background = pygame.sprite.Group()
background.add(Background())

all_sprites = pygame.sprite.Group()
all_sprites.add(NewShape())


pygame.init()

def collision():
    if pygame.sprite.spritecollide(shape.sprite, shapes, False):
        return True
    return False


def loop():
    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False

        pressed_keys = pygame.key.get_pressed()
        display.fill((0, 0, 0))
        [pygame.draw.rect(display, (40, 40, 40), i_rect, 1) for i_rect in grid]
        shape.update(pressed_keys)
        shape.draw(display)
        shapes.draw(display)
        pygame.display.update()
        clock.tick(60)
    pygame.quit()

loop()
